First stab at bumping the project to tokio 0.2

udoprog opened the pull request:
Sorry, this is a big change. But that's kinda expected :/

Related #182 

Non-exhaustive list of changes:
* Bumped edition to 2018.
* Dropped the use of `ClientExt` in favor if tiny bit of duplication in `ClientState` and macros.
* Removed `Client` trait and renamed `IrcClient` to `Client`.
* `PackedIrcClient` no longer exists.
* Renamed `error::IrcError` to `error::Error`.
* Removed all things associated with the reactor.
* `ClientStream` drives the stream for outgoing messages by default, unless it has been taken from the `Client` through `Client::outgoing`.
* You can access a thread-safe sender through `Client::sender`, which uses the outgoing queue for sending messages.
* Switched mutexes and locks to use [`parking_lot`](https://github.com/Amanieu/parking_lot).
* Updated all examples to use `#[tokio::main]` and tests to use `#[tokio::test]`.

#### TODO
* Commented out a test, since it requires more complex interaction than what is currently usable.
* ~~Commented out configuration tests since I couldn't run them locally (Windows dev environment).~~
* Burst-based rate limiter has not been ported yet.

udoprog committed:
Port to tokio 0.2

udoprog commented:
Removed the tests for reading `yaml`, `toml`, and `json` files from the filesystem because they complicated the build process for not a ton of benefit. For reference I'm building on Windows and running the shell script to generate the various representations is not straight forward.

serde's derives already do most of the heavy lifting. The one important part is probably just to check that the specific file formats remain unchanged in a separate serialize/deserialize test which doesn't use the filesystem.

aatxe commented:
It seems like you've also deleted a bunch of code having to do with saving configurations out, and the configuration format conversion application we had. Why?

udoprog commented:
I think it's because I dropped the tests as per https://github.com/aatxe/irc/pull/184#issuecomment-536191296 - I might've been too overzealous with deleting stuff as well so if you can tell me what I'll resurrect it. But I'll add tests which tests each format separately with files embedded in the test instead.

Do you still want to resurrect the config conversion utility in that case?

aatxe commented:
Yeah, I'm fine with killing the tests in favor of something that works more easily cross-platform, but I'd like to keep around the config conversion utility since it's just a useful quality of life tool if you decide to switch formats (and it's not like it has a very large maintenance burden associated with it).

udoprog commented:
Updated to futures-preview `0.3.0-alpha.19` and `tokio-0.2.0-alpha.6`. Also removed the direct dependency to `futures` in favor of more granular dependencies, all though `futures-util-preview` is still a big one.

We also now build on **beta**.

udoprog commented:
So the most up-to-date branch that I'm working on is **next**, and I intend to try and incorporate these changes once this has been accepted.
https://github.com/udoprog/irc/tree/next

Most notably, I've bumped the `irc-proto` crate to the 2018 edition and switch error handling to [`thiserror`](https://docs.rs/thiserror) since it provides implementations of `std::error::Error` which is much more future compatible.

freddyb commented:
Is this pull request still active?

udoprog commented:
@freddyb I'm still using it. But I'll bump it to the released tokio when I have the time. Still waiting for feedback from @aatxe though.

aatxe commented on src/client/mod.rs:
```diff
@@ -571,199 +788,243 @@ impl ClientState {
+    fn is_terminated(&self) -> bool {
+        // NB: outgoing stream never terminates.
+        // TODO: should it terminate if rx_outgoing is terminated?
+        false
```
My impression is that the answer to this should be yes. It’d be ideal if sending a QUIT message in IRC actually ended the program execution if the client wasn’t set up to restart in a loop or something of that sort.

udoprog replied:
Hm. Ending the execution is simply a matter of dropping the client and stream. I don't think QUIT messages are anything but informative, and the real way to quit a server is to disconnect. But I could be wrong.

mozfreddyb replied:
This is the client though? :thinking:

aatxe commented on src/client/mod.rs:
```diff
@@ -979,164 +1229,176 @@ mod test {
 
-    #[test]
+    // TODO: this seems to require a somewhat more complex interaction with the
+    // client.
```
I don’t follow. This is the same as the next test except that it involves sending a PART.

udoprog replied:
I'll have to comment it out and test it. I don't remember why it didn't work. Will have a look when I port it to stable tokio.

aatxe commented on src/client/transport.rs:
```diff
@@ -1,241 +1,208 @@
-            AsyncSink::NotReady(item) => Ok(AsyncSink::NotReady(item)),
-            AsyncSink::Ready => {
-                self.current_burst_messages += 1;
-                self.rolling_burst_window.push_back(self.burst_timer.sleep(Duration::from_secs(
```
Why did you completely wipe out all the flood control settings?

udoprog replied:
Ah. Yeah, it hasn't been ported yet (see TODO above).

I also wanted to pitch not including it since plugging in an external rate limiter is fairly straight forward. For example, I've written and use https://docs.rs/leaky-bucket/0.6.0/leaky_bucket/

Alternatively if this can be put into a higher level abstraction so me as a user can decide whether or not to include it, if you want to have access to just the protocol bits.

Either way, I'm not sure I have time to port it right now. But you can always delay a release until it has been added again.

udoprog commented:
Bumped the branch to the stable tokio 0.2 now.

udoprog commented:
@aatxe any chance to get this merged? Most of the things are in place, and the things which are incomplete can be fixed incrementally in master before a release. Having this hanging for so long is somewhat taxing.
